### PR TITLE
Upgrade to jupyterlab4 with lab4_port commits

### PR DESCRIPTION
**Pull request by oc-will-holtam**
I've taken my changes from [here](https://github.com/jupyterlab/jupyterlab_pygments/pull/36), which pass the various github action checks.
And cherry-picked the relevant un-merged changes from [here](https://github.com/jupyterlab/jupyterlab_pygments/pull/31).

### CODE REVIEW

**Comment by WilliamHoltam** [comment]
@krassowski ready for review & added PR label :)

**Comment by benz0li** [comment]
> @krassowski ready for review & added PR label :)

Does it also build with Python 3.12?

File: package-lock.json
Comment (krassowski) [inline package-lock.json:1]: I think `yarn.lock` should be enough? Certainly we would not want to have two lockfiles.

Comment (WilliamHoltam) [reply package-lock.json:1]: ok, thanks, I removed it

**Comment by WilliamHoltam** [comment]
> Does it also build with Python 3.12?

It does now :)

**Comment by benz0li** [comment]
> Should this PR be rebooted using the extension upgrade script? https://jupyterlab.readthedocs.io/en/latest/extension/extension_migration.html#upgrading-extension-using-the-upgrade-script
> 
> This would help pick up the latest changes to the extension template.

– https://github.com/jupyterlab/jupyterlab_pygments/pull/31#issuecomment-1681699291

@jtpio Does this meet the expectations? Are you responsible for reviewing and merging?

**Comment by krassowski** [comment]
Can you update jupyterlab/python/node versions in `binder/environment.yml` to make testing easier?

**Comment by WilliamHoltam** [comment]
> Can you update jupyterlab/python/node versions in `binder/environment.yml` to make testing easier?

Which versions do you need, the latest of each?

**Comment by WilliamHoltam** [comment]
Please add comments if there are any further changes I can make to make things easier for you :)

**Review by krassowski — APPROVED** [review]
Thank you @WilliamHoltam (and @RRosio) this looks good to me. I will wait with merging until Monday so that @jtpio has a chance to take a look.

**Review by jtpio — APPROVED** [review]
Diff looks good, thanks!

**Comment by benz0li** [comment]
@krassowski It would be great to have this released by the end of the week.

Thanks to everyone involved!

**Comment by benz0li** [comment]
@martinRenou Thank you for releasing v0.3.0.

**Pull request merged**